### PR TITLE
anthropic[major]: support python 3.13

### PR DESCRIPTION
**Pull request by ccurme**
Last week Anthropic released version 0.39.0 of its python sdk, which enabled support for Python 3.13. This release deleted a legacy `client.count_tokens` method, which we currently access during init of the `Anthropic` LLM. Anthropic has replaced this functionality with the [client.beta.messages.count_tokens() API](https://github.com/anthropics/anthropic-sdk-python/pull/726).

To enable support for `anthropic >= 0.39.0` and Python 3.13, here we drop support for the legacy token counting method, and add support for the new method via `ChatAnthropic.get_num_tokens_from_messages`.

To fully support the token counting API, we update the signature of `get_num_tokens_from_message` to accept tools everywhere.

### CODE REVIEW

**Comment by ccurme** [comment]
Integration tests pass with 3.13 https://github.com/langchain-ai/langchain/actions/runs/11688847362

File: libs/partners/anthropic/langchain_anthropic/llms.py [hunk 1]
```diff
@@ -375,9 +374,11 @@ async def _astream(
 
     def get_num_tokens(self, text: str) -> int:
         """Calculate number of tokens."""
-        if not self.count_tokens:
-            raise NameError("Please ensure the anthropic package is loaded")
-        return self.count_tokens(text)
+        raise NotImplementedError(
```
Comment (efriis) [inline libs/partners/anthropic/langchain_anthropic/llms.py:377]: to make this nonbreaking could we just call the new endpoint with a single humanmessage and subtract a constant number of standard tokens. (for e.g. `Human: `, and print a warning?

Comment (ccurme) [reply libs/partners/anthropic/langchain_anthropic/llms.py:377]: Older models like claude 2 not supported unfortunately: https://docs.anthropic.com/en/docs/build-with-claude/token-counting#supported-models

File: libs/partners/anthropic/langchain_anthropic/chat_models.py [hunk 2]
```diff
@@ -1113,6 +1113,40 @@ class AnswerWithJustification(BaseModel):
         else:
             return llm | output_parser
 
+    @beta()
+    def get_num_tokens_from_messages(self, messages: List[BaseMessage]) -> int:
```
Comment (baskaryan) [inline libs/partners/anthropic/langchain_anthropic/chat_models.py:1117]: thoughts on updating base implementation to take tools and/or kwargs? definitely not an anthropic-specific problem

Comment (efriis) [reply libs/partners/anthropic/langchain_anthropic/chat_models.py:1117]: id support

File: libs/core/langchain_core/language_models/base.py [hunk 3]
```diff
@@ -364,13 +364,22 @@ def get_num_tokens(self, text: str) -> int:
         """
         return len(self.get_token_ids(text))
 
-    def get_num_tokens_from_messages(self, messages: list[BaseMessage]) -> int:
+    def get_num_tokens_from_messages(
```
Comment (baskaryan) [inline libs/core/langchain_core/language_models/base.py:367]: should we emit a warning here

File: libs/partners/openai/langchain_openai/chat_models/base.py [hunk 4]
```diff
@@ -897,7 +902,16 @@ def get_num_tokens_from_messages(self, messages: List[BaseMessage]) -> int:
...
+        main/examples/How_to_format_inputs_to_ChatGPT_models.ipynb
+
+        Args:
+            messages: The message inputs to tokenize.
+            tools: If provided, sequence of dict, BaseModel, function, or BaseTools
+                to be converted to tool schemas.
+        """
+        # TODO: Count bound tools as part of input.
+        if tools is not None:
+            warnings.warn("Counting tokens in tool schemas is not yet supported.")
```
Comment (baskaryan) [inline libs/partners/openai/langchain_openai/chat_models/base.py:914]: could we make it explicit that we're ignore tools
```suggestion
            warnings.warn("Counting tokens in tool schemas is not yet supported. Ignoring tools.")
```

**Pull request merged**